test/model/binding.dart

```diff
@@ -87,9 +93,37 @@ class TestZulipBinding extends ZulipBinding {
           ),
         ]);
       }
-      _debugAlreadyLoaded = true;
+      _debugAlreadyLoadedStore = true;
       return true;
     }());
     return Future.value(globalStore);
   }
+
+  /// The value that `ZulipBinding.instance.launchUrl()` should return.
+  ///
+  /// See also [takeLaunchUrlCalls].
+  bool launchUrlResult = true;
+
+  /// Consume the log of calls made to `ZulipBinding.instance.launchUrl()`.
+  ///
+  /// This returns a list of the arguments to all calls made
+  /// to `ZulipBinding.instance.launchUrl()` since the last call to
+  /// either this method or [reset].
+  ///
+  /// See also [launchUrlResult].
+  List<({Uri url, url_launcher.LaunchMode mode})> takeLaunchUrlCalls() {
+    final result = _launchUrlCalls;
+    _launchUrlCalls = null;
+    return result ?? [];
+  }
+  List<({Uri url, url_launcher.LaunchMode mode})>? _launchUrlCalls;
+
+  @override
+  Future<bool> launchUrl(
+    Uri url, {
+    url_launcher.LaunchMode mode = url_launcher.LaunchMode.platformDefault,
+  }) async {
+    (_launchUrlCalls ??= []).add((url: url, mode: mode));
+    return launchUrlResult;
+  }
```

Comment:
Cool! One obstacle that's been in the way of testing the compose box's upload-and-attach buttons is that they use Flutter plugins. This pattern seems helpful for addressing that, too.

Reply:
Yeah, definitely.

If we did end up splitting the bindings into several mixin classes for code organization, I'm imagining the split would look like
* data
* notifications
* miscellaneous plugins that have one or two relevant methods each, where the plugins behind those compose-box buttons would be a prime example (and `url_launcher` is another prime example).